chainer/functions/normalization/local_response_normalization.py

```diff
@@ -54,86 +54,133 @@ def check_type_forward(self, in_types):
             x_type.ndim >= 2,
         )
 
-    def forward_cpu(self, x):
+    def forward_cpu(self, inputs):
         if (intel64.should_use_ideep('>=auto')
-                and intel64.inputs_all_ready(x, (4,))):
+                and intel64.inputs_all_ready(inputs, (4,))):
             self._use_ideep = True
-            return self._forward_ideep(x)
+            return self.forward_ideep(inputs)
+
+        x, = inputs
+        self.retain_inputs((0,))
+        self.retain_outputs((0,))
 
         half_n = self.n // 2
-        x2 = numpy.square(x[0])
+        x2 = numpy.square(x)
         sum_part = x2.copy()
         for i in six.moves.range(1, half_n + 1):
             sum_part[:, i:] += x2[:, :-i]
             sum_part[:, :-i] += x2[:, i:]
         self.unit_scale = self.k + self.alpha * sum_part
         self.scale = self.unit_scale ** -self.beta
-        self.y = x[0] * self.scale
-        return self.y,
+        y = x * self.scale
+        return y,
+
+    def forward_ideep(self, inputs):
+        x, = inputs
+        self.retain_inputs((0,))
+        self.retain_outputs((0,))
 
-    def _forward_ideep(self, x):
         param = intel64.ideep.localResponseNormalizationParam(
             self.n, self.k, self.n * self.alpha, self.beta,
             intel64.ideep.localResponseNormalizationParam.lrn_across_channels)
         y, indexes = intel64.ideep.localResponseNormalization.Forward(
-            intel64.ideep.array(x[0]), param)
-        self.y = y
+            intel64.ideep.array(x), param)
         self.indexes = indexes
+        return y,
+
+    def forward_gpu(self, inputs):
+        x, = inputs
+        self.retain_inputs((0,))
+        self.retain_outputs((0,))
+
+        self.y = cuda.cupy.square(x)  # temporary
+        self.scale = cuda.cupy.empty_like(self.y)
+        _cu_conv_sum(self.scale, self.y, self.n)
+        cuda.elementwise(
+            'T x, T k, T alpha, T beta',
+            'T y, T scale',
+            '''scale = k + alpha * scale;
+               y = x * pow(scale, -beta);''',
+            'lrn_fwd')(x, self.k, self.alpha, self.beta,
+                       self.y, self.scale)
         return self.y,
 
-    def backward_cpu(self, x, gy):
+    def backward(self, indexes, grad_outputs):
+        x, = self.get_retained_inputs()
+        y, = self.get_retained_outputs()
+        gy, = grad_outputs
+
+        scale = getattr(self, 'scale', None)
```

Comment:
Would you initialize these instance variables with `None` in `__init__()` for efficiency?

Reply:
Thank you for quick review. I fixed this point.